src/main/scala/fastq.scala

```diff
@@ -165,91 +165,91 @@ case object fastq {
+      def rec(cs: String, acc: collection.mutable.Builder[Int,Vector[Int]], errors: Boolean): Option[Quality] =
         if(errors) { None } else {
 
-          if(cs.isEmpty) Some( Quality(acc) ) else {
+          if(cs.isEmpty) Some( Quality(acc.result) ) else {
 
             cs.head.toInt match {
-              case q if 33 <= q && q <= 126 => rec(cs.tail, acc :+ (q - 33), errors)
+              case q if 33 <= q && q <= 126 => rec(cs.tail, acc += (q - 33), errors)
               case _                        => rec(cs, acc, errors = true)
             }
           }
         }
 
-      rec(raw, Seq.empty, false)
+      val bldr = Vector.newBuilder[Int]; bldr.sizeHint(raw.length)
+
+      rec(raw, bldr, false)
     }
 
     val toPhred33: Int => Char = { i => (i + 33).toChar }
   }
 
-  class Id private[fastarious] (val value: String) extends AnyVal {
+  case class Id private[fastarious] (val value: String) extends AnyVal {
 
     def asString: String =
       s"@${value}"
+
+    def isEmpty: Boolean =
+      value.isEmpty
   }
 
   case object Id {
 
-    private val isValid: String => Boolean = { raw =>
-      (raw startsWith "@") &&
-      !(raw contains "\n")
-    }
-
-    def from(raw: String): Option[Id] = {
+    def from(raw: String): Id =
+      new Id(raw.filterNot(_ == '\n'))
 
-      // just *one* line
-      // FIXME: raw starts with several @, it will drop them all
-      val l = raw.filterNot(_ == '\n').dropWhile(_ == '@')
+    /*
+      ### Id parsing
 
-      if(l.nonEmpty) Some( new Id(l) ) else None
-    }
+      These methods are *not* for building Id values; for that use `from`.
+    */
+    private val isValid: String => Boolean =
+      _ startsWith "@"
 
-    // FIXME: this passes the raw string (with @) to the constructor
+    /*
+      `parseFrom` will drop the *first* '@' in `raw`, if any. My understanding is that `@@hola` is a valid (albeit confusing) FASTQ id.
+    */
     def parseFrom(raw: String): Option[Id] =
-      if(isValid(raw)) Some( new Id(raw) ) else None
+      if(isValid(raw)) Some( new Id(raw.stripPrefix("@")) ) else None
   }
 
-  case class FASTQ(val id: Id, val sequence: Sequence) {
+  case class FASTQ(val id: Id, val value: Sequence) {
 
     def asStringPhred33: String = Seq(
       id.asString,
-      sequence.asStringPhred33
+      value.asStringPhred33
     ).mkString("\n")
 
     def toFASTA: FASTA.Value = FASTA(
       fasta.header( FastaHeader(id.value) )              ::
-      fasta.sequence( FastaSequence(sequence.sequence) ) ::
+      fasta.sequence( FastaSequence(value.sequence) ) ::
       *[AnyDenotation]
     )
   }
 
   case object FASTQ {
 
     def fromStringsPhred33(
-      id: String,
-      sequence: String,
-      quality: String
-    ): Option[FASTQ] = {
-
-      for {
-        i <- Id.from(id)
-        s <- Sequence.fromStringsPhred33(sequence, quality)
-      } yield FASTQ(i, s)
-    }
+      id        : String,
```

Comment:
I just meant that when I take this method, I don't know what's the supposed input for it: a cleaned up ID or a raw line (that I got from somewhere else)